### PR TITLE
feat(perf-tests): Add Performance tests and run them regularly in release pipeline

### PR DESCRIPTION
**Pull request by adambollen**
### Description

This adds perf tests under a separate test category that run a set of queries from a static file, measures the "non-query time" (i.e. driver overhead), and outputs the `P50/P95/P99/StdDev` for the data points in a text file to be sent to a Slack channel or whatever.  Support for shipping to Datadog has also been added, but is not currently being used.

Each perf test in the json file has the following pieces:

- A name for the test scenario
- Query parts to be used in query composition
- Whether the result is from the `Product` collection from the FSL file (and modeled in `TestDataModels.cs`)
  - If true, the test should deserialize the return object(s)
- Whether the result returns multiple pages of data
  - If true, the test will paginate through all the results and measure the overall time

### Motivation and context

Initially, we want to have some data on the performance characteristics of the driver as a whole to see if we're taking longer than expected in some scenarios, thus this change will run them multiple times per day via the release pipeline.  In the future, this can also help to identify perf regressions in a PR.

### How was the change tested?

The tests were run extensively locally, as well as setting up a Docker container the same way the Concourse pipeline will and ensuring that the tests continue to run in that environment.

### Screenshots (if appropriate):
The Slack message at the end of the pipeline tests looks like this:
image

### Change types

* - [ ] Bug fix (non-breaking change that fixes an issue)
* - [x] New feature (non-breaking change that adds functionality)
* - [ ] Breaking change (backwards-incompatible fix or feature)

### Checklist:

* - [x] My code follows the code style of this project.
* - [ ] My change requires a change to Fauna documentation.
* - [ ] My change requires a change to the README, and I have updated it accordingly.

----
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Fauna.Test/Performance/TestQueryExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+{
+    /// <summary>
+    /// Get a <see cref="Query"/> composited from the given string query fragments
+    /// </summary>
+    /// <param name="queryParts">List of strings from which to create the query</param>
+    /// <returns>A <see cref="QueryExpr"/> representing the composite query</returns>
+    public static Query GetCompositedQueryFromParts(this List<string> queryParts)
+    {
+        return new QueryExpr(queryParts.Select(x => new QueryLiteral(x)).ToArray());
+    }
```
Comment (pnwpedro) [inline Fauna.Test/Performance/TestQueryExtensions.cs:16]: I think we cover one more important piece of this lib if we use the QueryStringHandler directly. Here's an example of how to do that. Pretty sure we can dream up a way to configure variables via the queries.json as well–reflection and Activator and such–although that can be a follow on if necessary.

```
        var handler = new QueryStringHandler(0,0);
        handler.AppendLiteral("let m = ");
        handler.AppendFormatted(42);
        handler.AppendLiteral("; m");
        var q = FQL(ref handler);
        var r = _client.QueryAsync<int>(q);
```

**Review by pnwpedro — CHANGES_REQUESTED** [review]

**Comment by jdcryans** [comment]
Overall looks good. I'm very interested by the results. Looks like 20ish ms is the floor for overhead? That seems excessive.

File: Fauna.Test/Performance/MetricsHandler.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+/// <summary>
+/// Provides an abstraction layer to track perf test metrics either to a
+/// local CSV file or to Datadog
+/// </summary>
+internal class MetricsHandler
+{
+    private const string RawStatsFilename = "rawstats.csv";
+    private const string StatsBlockFilename = "stats.txt";
```
[the remark this answers is not in the document]
Comment (pnwpedro) [reply Fauna.Test/Performance/MetricsHandler.cs:30]: Yeah, exactly this. A timestamp should be good enough.

**Review by pnwpedro — APPROVED** [review]
It looks like we might want to do some perf profiling to see where we're spending cycles!

**Review by cynicaljoy — APPROVED** [review]

**Pull request merged**